src/subxt_transaction.rs

```diff
@@ -496,10 +505,107 @@ pub(crate) fn build_eth_tx_payload(
 	}
 }
 
+#[allow(clippy::too_many_arguments)]
+async fn create_online_transaction<C: subxt::Config, KP, G>(
+	from_keypair: &KP,
+	nonce: u128,
+	mortality: &Option<u64>,
+	account: &str,
+	sink: &SubxtTransactionsSink<C, KP>,
+	from_account_id: &<C as subxt::Config>::AccountId,
+	to_account_id: &<C as subxt::Config>::AccountId,
+	recipe: &TransactionRecipe,
+	generate_payload: G,
+) -> Result<SubxtTransaction<C>, Error>
+where
+	G: GenerateTxPayloadFunction<AccountIdOf<C>>,
+	AccountIdOf<C>: Send + Sync + AsRef<[u8]>,
+	KP: Signer<C> + Clone + Send + Sync + 'static,
+	<<C as subxt::Config>::ExtrinsicParams as subxt::config::ExtrinsicParams<C>>::Params: From<(
+		(),
+		(),
+		(),
+		CheckNonceParams,
+		(),
+		CheckMortalityParams<C>,
+		ChargeAssetTxPaymentParams<C>,
+		ChargeTransactionPaymentParams,
+		(),
+	)>,
+{
+	// Needed because `Params` as associated type does not implement clone, and we need to
+	// recreate the tx params in a loop when we can't create a partial tx with the online
+	// client, due to various RPC related issues or state not being up to date (currently we
+	// handle an error which happens when trying to create a partial tx that is based on a
+	// certain finalized block returned by the RPC, which is then reported as not found).
+	// Retrying seems to fix the issue.
+	fn tx_params<CC: subxt::Config>(
+		mortality: &Option<u64>,
+		nonce: u64,
+		recipe: &TransactionRecipe,
+	) -> <DefaultExtrinsicParams<CC> as ExtrinsicParams<CC>>::Params {
+		let mut params = <SubstrateExtrinsicParamsBuilder<CC>>::new().nonce(nonce).tip(recipe.tip);
+		if let Some(mortal) = mortality {
+			params = params.mortal(*mortal);
+		}
+		params.build()
+	}
+
+	async fn subxt_transaction<CC: subxt::Config, KEYP>(
+		sink: &SubxtTransactionsSink<CC, KEYP>,
+		mut partial_tx: PartialTransaction<CC, OnlineClient<CC>>,
+		from_keypair: &KEYP,
+		nonce: u128,
+		mortality: &Option<u64>,
+		account: &str,
+	) -> Result<SubxtTransaction<CC>, Error>
+	where
+		KEYP: Signer<CC> + Clone + Send + Sync + 'static,
+		AccountIdOf<CC>: Send + Sync + AsRef<[u8]>,
+	{
+		let block_ref = sink
+			.api()
+			.backend()
+			.latest_finalized_block_ref()
+			.await
+			.expect("to get the last finalized block ref. qed");
+		let block = sink
+			.api()
+			.blocks()
+			.at(block_ref)
+			.await
+			.expect("to get the corresponding block header. qed");
+		let block_number = block.number().into();
```

Comment:
This deserves some comment, as it is tricky and the logic is aligned with subxt:
https://github.com/paritytech/subxt/blob/77b6abccbacf194f3889610024e2f4024e8c2822/subxt/src/tx/tx_client.rs#L600

Would be nice to see a reference to `inject_account_nonce_and_block` function.

I think that it also allows for some little race condition - if some block is finalized in meantime. But don't see a better way to implement it now. Maybe we could ask subxt if they could expose actual blocks in `Params/DefaultExtrinsicParams` somehow?

Reply:
> I think that it also allows for some little race condition - if some block is finalized in meantime. But don't see a better way to implement it now. 

Yes, but thought that even if a new finalized block happens in the meantime, that means will wait just a little longer with the block monitor for the transaction to be finalized (which is not that bad I think).

> Would be nice to see a reference to inject_account_nonce_and_block function.

Done here: [2df3030](https://github.com/michalkucharczyk/tx-test-tool/pull/43/commits/2df3030031cd9f719f7ceb19ccb00856151bbd68)

Reply:
> Maybe we could ask subxt if they could expose actual blocks in Params/DefaultExtrinsicParams somehow?

Yes, to be 100% accurate with this. I'll file a question.

Reply:
I am reconsidering this a bit. I think we'll need either way some online capability to fetch last finalized block. Can leave this unresolved for now until we'll tackle #44 . We should due some preliminary research then about options before asking for things.